Products/PerFactErrors/errors.py

```diff
@@ -50,7 +51,53 @@ def afterfail_error_message(event):
         if body is not None:
             req.response.setBody(body)
     except Exception:
+        logger.warn(traceback.format_exc())
         logger.exception('Error while rendering error message')
         transaction.abort()
     else:  # no exception
         transaction.commit()
+
+
+class PerFactException(Exception):
+    ''' Special Exception Class
+    '''
+
+    def __init__(self, msg='', show_to_user=False,
+                 apperrorlog=False, payload=None, **kw):
+        '''
+        Input:
+        - "msg" (string) is the error message,
+          possibly containing placeholders like "{lotnumber}"
+        - "show_to_user" (boolean, default False), controls
+          if the error is intended to be shown to the end
+          user even on a production system. Also causes the
+          error to be translated before showing it to the user
+          (not the one that is inserted into the database). If
+          it is not set, the user simply gets a generic "An
+          error occured" with an ID.
+        - "apperrorlog" (boolean, default True), controls if
+          the error is logged in the apperrorlog
+        - "args" (dictionary, default None which means {})
+          contains values that should be inserted into the
+          placeholders, possibly after translation.
+        - **kw should be used to allow more arguments to be
+          passed later. Anything given here should be stored in
+          the class instance.
+        '''
+
+        super(PerFactException, self).__init__()
+        self.msg = msg
+        self.show_to_user = show_to_user
+        self.apperrorlog = apperrorlog
+        self.payload = payload or {}
+
+    def __str__(self):
+        return repr((self.msg, self.payload))
```

Comment:
Sorry I did not understand why this is necessary. The error handling should not stringify instances of `PerFactException`. But I guess I will see that when reviewing the changes in the Data.FS. By itself and for debugging this does not hurt.

Reply:
Without the function overloading, no error message was printed. The only information you got in the log was the exception type.